src/modules/CSADigitizer/CSADigitizerModule.cpp

```diff
@@ -0,0 +1,337 @@
+    // inverse Laplace transform R/((1+a*s)*(1+s*b)) is (wolfram alpha) (R (e^(-t/a) - e^(-t/b))) /(a - b) 
+    fImpulseResponse_ = new TF1("fImpulseResponse_", "[0] * ( exp(-x / [1]) - exp(-x/[2]) ) / ([1]-[2])", 0, tmax_);
+    fImpulseResponse_->SetParameters(rf_, tauF_, tauR_);
+
+
+}
+
+
+
+void CSADigitizerModule::init() {
+
+    if(config_.get<bool>("output_plots")) {
+        LOG(TRACE) << "Creating output plots";
+
+        // Plot axis are in kilo electrons - convert from framework units!
+        int maximum = static_cast<int>(Units::convert(config_.get<int>("output_plots_scale"), "ke"));
+        auto nbins = config_.get<int>("output_plots_bins");
+
+	//asv do these histos make sense? what about one with the noise?
+        // Create histograms if needed
+        h_pxq = new TH1D("pixelcharge", "raw pixel charge;pixel charge [ke];pixels", nbins, 0, maximum);
+        h_tot = new TH1D("tot", "time over threshold;time over threshold [ns];pixels", nbins, 0, maximum);
+        h_toa = new TH1D("toa", "time of arrival;time of arrival [ns];pixels", nbins, 0, maximum);
+	h_pxq_vs_tot = new TH2D("pxqvstot", "ToT vs raw pixel charge;pixel charge [ke];ToT [ns]", nbins, 0, maximum, nbins, 0, tmax_);
+
+    }
+
+    
+}
+
+
+
+void CSADigitizerModule::run(unsigned int event_num) {
+    // Loop through all pixels with charges
+    std::vector<PixelHit> hits;
+    for(auto& pixel_charge : pixel_message_->getData()) {
+        auto pixel = pixel_charge.getPixel();
+        auto pixel_index = pixel.getIndex();
+        auto inputcharge = static_cast<double>(pixel_charge.getCharge());
+
+        LOG(DEBUG) << "Received pixel " << pixel_index << ", charge " << Units::display(inputcharge, "e");
+
+	const auto& pulse = pixel_charge.getPulse(); // the pulse containing charges and times
+	auto pulse_vec = pulse.getPulse();           // the vector of the charges 
+        auto timestep = pulse.getBinning();
+	const int npx = static_cast<int>(ceil(tmax_/timestep));
+
+	if (first_event_){ // initialize impulse response function - assume all time bins are equal
+	  impulseResponse_ = new double[npx];
+	  for (int ipx=0; ipx<npx; ++ipx){
+	    impulseResponse_[ipx]=fImpulseResponse_->Eval(ipx*timestep);
+	  }
+	  first_event_ = false;
+	  LOG(TRACE) << "impulse response initalised. timestep  : " <<  timestep << ", tmax_ : " <<  tmax_ << ", npx " << npx;
+	}
+	
+	Pulse output_pulse(timestep);
+	auto input_length = pulse_vec.size();
+	LOG(TRACE) << "Preparing pulse for pixel " << pixel_index << ", " << pulse_vec.size() << " bins of "
+		   << Units::display(timestep, {"ps", "ns"})
+		   << ", total charge: " << Units::display(pulse.getCharge(), "e");
+	//convolution of the pulse (size input_length) with the impulse response (size npx)
+	for(unsigned long int k=0; k<npx; ++k){
+	  for(unsigned long int i=0; i<=k; ++i){
+	    if( (k-i) < input_length){
+	      //asv to do: not a charge, but voltage pulse... still use this?
```

Comment:
Since we are not storing this pulse anywhere I would suggest not using it but simply store the time steps in some other container just for plotting purposes.

Reply:
Hm, it actually might even make sense to use `Pulse`, let me think about this one a bit more

Reply:
New idea: what if I add "addVoltage" and "getVoltage" to `Pulse` and use that instead of add/getCharge?

Reply:
I thin it would be easier and less "abusive" of objects if we simply use a vector here and append. The content will not be stored long-term anyway but just used for calculating ADC/ToT/ToA.

If you add `setVoltage`/`getVoltage` mehtods, you also need to make sure that you cannot call `setVoltage` and `getCharge` together - which in the end make it two completely disjunct objects.

Reply:
true, if the object never leaves this module, there is no need for the extra features like "getBinning"... I've made it a vector now.